File: easy/third_max/third_max.py
class Solution(object):
    def thirdMax_first_solution(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        first_max = second_max = third_max = -float("inf")
        for val in nums:
            if val in (first_max, second_max, third_max):
                continue
            elif val > first_max:
                third_max, second_max, first_max = second_max, first_max, val
            elif val > second_max:
                third_max, second_max = second_max, val
            elif val > third_max:
                third_max = val

        if third_max > -float("inf"):
            return third_max

        return first_max

    def third_max_second_solution(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        first_max = max(nums)
        nums = list(filter(lambda x: x != first_max, nums))

        if len(nums) != 0:
            second_max = max(nums)
            nums = list(filter(lambda x: x != second_max, nums))

            if len(nums) != 0:
                third_max = max(nums)
                return third_max

        return first_max


s = Solution()
print("Solution 1 : ", s.thirdMax_first_solution([3, 2, 1]))
print("Solution 2 : ", s.thirdMax_first_solution([1, 2]))
print("Solution 3 : ", s.thirdMax_first_solution([2, 2, 3, 1]))

File: easy/array/third_max/third_max.go
package main

import (
	"fmt"
	"math"
)

func thirdMax(nums []int) int {
	firstMax, secondMax, thirdMax := math.MinInt64, math.MinInt64, math.MinInt64
	for i := 0; i < len(nums); i++ {
		if nums[i] > firstMax {
			thirdMax = secondMax
			secondMax = firstMax
			firstMax = nums[i]
		} else if nums[i] > secondMax && nums[i] < firstMax {
			thirdMax = secondMax
			secondMax = nums[i]
		} else if nums[i] > thirdMax && nums[i] < secondMax {
			thirdMax = nums[i]
		}
	}

	if thirdMax > math.MinInt64 {
		return thirdMax
	}

	return firstMax
}

func main() {
	fmt.Println("Solution 1 : ", thirdMax([]int{3, 2, 1}))
	fmt.Println("Solution 2 : ", thirdMax([]int{1, 2}))
	fmt.Println("Solution 3 : ", thirdMax([]int{2, 2, 3, 1}))
}
